### PR TITLE
Add Dissect Processor

### LINKED ISSUE #3362: Dissect processor
**Is your feature request related to a problem? Please describe.**
A dissect processor (similar to https://www.elastic.co/guide/en/logstash/current/plugins-filters-dissect.html#plugins-filters-dissect) would be useful when extracting fields from log messages that follows a repeated pattern. It's simpler to use and can be faster than grok in certain use cases.

**Describe the solution you'd like**

**Describe alternatives you've considered (Optional)**

**Additional context**

### PR DESCRIPTION
**Pull request by oeyh**
### Description
Moved commits from #3090 by @vishalboin, rebased to resolve merge conflicts and made some small tweaks on tests and docs.
 
### Issues Resolved
Resolves #3362
 
### Check List
- [x] New functionality includes testing.
- [ ] New functionality has a documentation issue. Please link to it in this PR.
  - [ ] New functionality has javadoc added
- [x] Commits are signed with a real name per the DCO

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/data-prepper/blob/main/CONTRIBUTING.md).

### CODE REVIEW

File: data-prepper-plugins/dissect-processor/src/main/java/org/opensearch/dataprepper/plugins/processor/dissect/Fields/FieldHelper.java [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public class FieldHelper {
```
Comment (graytaylor0) [inline data-prepper-plugins/dissect-processor/src/main/java/org/opensearch/dataprepper/plugins/processor/dissect/Fields/FieldHelper.java:17]: Looks like some missing unit tests on this class.

File: data-prepper-plugins/dissect-processor/src/main/java/org/opensearch/dataprepper/plugins/processor/dissect/Dissector.java [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+
+public class Dissector {
```
Comment (graytaylor0) [inline data-prepper-plugins/dissect-processor/src/main/java/org/opensearch/dataprepper/plugins/processor/dissect/Dissector.java:25]: Looks like this class has no unit tests and is being tested indirectly in `DissectProcessorTest`. Ideally we would mock this class and have unit tests for it directly, but I won't block the PR on that

File: data-prepper-plugins/dissect-processor/src/main/java/org/opensearch/dataprepper/plugins/processor/dissect/Fields/FieldHelper.java (hunk 1 above)
Comment (oeyh) [reply data-prepper-plugins/dissect-processor/src/main/java/org/opensearch/dataprepper/plugins/processor/dissect/Fields/FieldHelper.java:17]: Added.

File: data-prepper-plugins/dissect-processor/src/main/java/org/opensearch/dataprepper/plugins/processor/dissect/Dissector.java (hunk 2 above)
Comment (oeyh) [reply data-prepper-plugins/dissect-processor/src/main/java/org/opensearch/dataprepper/plugins/processor/dissect/Dissector.java:25]: Makes sense. I separated the tests for Dissector from DissectProcessorTest. `DissectProcessorTest` now only test things after the actual dissection is done - e.g., handle dissect exception, convert type, and put dissected fields into event.

**Review by graytaylor0 — APPROVED** [review]

**Review by kkondaka — APPROVED** [review]
Looks good. I have looked at these changes when Vishal submitted PR#3090

**Pull request merged**